File: rules-interpreter/src/main/java/com/github/iamrenny/ruleflow/visitors/ActionsVisitor.java
package io.github.iamrenny.ruleflow.util;

import io.github.iamrenny.ruleflow.RuleFlowLanguageParser;
import io.github.iamrenny.ruleflow.utils.Pair;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class ActionsVisitor {

    public List<Pair<String, Map<String, String>>> visit(RuleFlowLanguageParser.ActionsContext context) {
        return context.action().stream()
            .map(action -> {
            String name;
            if (action.K_ACTION() != null) {
                name = action.param_value.getText().replace("'", "");
            } else if ("manual_review".equals(action.action_id.getText())) {
                name = action.action_id.getText().replace("'", "");
            } else if ("apply_restriction".equals(action.action_id.getText())) {
                name = action.action_id.getText().replace("'", "");
            } else if (action.action_id.getText() != null) {
                name = action.action_id.getText().replace("'", "");
            } else {
                throw new IllegalArgumentException("Cannot find action name or identifier in " + action.getText());
            }
            return new Pair<>(name, new ActionVisitor().visit(action));
        })
        .collect(Collectors.toList());
    }
}

class ActionVisitor {

    public Map<String, String> visit(RuleFlowLanguageParser.ActionContext action) {
        if (action.action_params() != null) {
            return action.action_params().param_pairs().param_pair().stream()
                .collect(Collectors.toMap(
                    a -> a.field_name.getText().replace("'", ""),
            a -> a.field_value.getText().replace("'", "")
            ));
        } else {
            return Map.of();
        }
    }
}

File: rules-interpreter/src/main/kotlin/com/rappi/fraud/rules/parser/visitors/ActionsVisitor.kt
package com.rappi.fraud.rules.parser.visitors

import com.rappi.fraud.analang.ANAParser
import kotlin.streams.toList

class ActionsVisitor {
    fun visit(context: ANAParser.ActionsContext): Map<String, Map<String, String>> {
        return context.action()
            .stream()
            .map { action ->
                val name = when {
                    action.K_ACTION() != null -> action.param_value.text.replace("'", "")
                    action.action_id.text == "manual_review" -> action.action_id.text.replace("'", "")
                    action.action_id.text == "apply_restriction" -> action.action_id.text.replace("'", "")
                    else -> error("Cannot find action name or identifier in ${action.text}")
                }
                Pair(name, ActionVisitor().visit(action))
            }
            .toList()
            .toMap()

    }
}

class ActionVisitor {

    fun visit(action: ANAParser.ActionContext): Map<String, String> {

        return if (action.action_params() != null)
            action.action_params().string_literal().asSequence().withIndex()
                .map { (i, node) -> IndexedValue<String>(i, node.text) }
                .map { (i, name) ->  IndexedValue(i, name.replace("'","")) }
                .map { (i, name) ->
                    Pair(name, action.action_params().validValue(i).text.replace("'", ""))
                }
                .toMap()
        else mapOf()
    }
}
